### PR TITLE
[REP] Ray accelerator support

### PR DESCRIPTION
**Pull request by jjyao**
How to support new accelerators in Ray end-to-end.

### CODE REVIEW

File: reps/2023-10-13-accelerator-support.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    ...
+```
+
+#### Implementation
+
+Core defines an `Accelerator` base class and each accelerator needs to implement a subclass (e.g. `TPUAccelerator`):
+
+```python
+@DeveloperAPI
+class Accelerator(ABC):
```
Comment (jjyao) [inline reps/2023-10-13-accelerator-support.md:113]: Implementation is here: https://github.com/ray-project/ray/pull/40286

File: reps/2023-10-13-accelerator-support.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+Users can use the corresponding accelerator resources for data operations:
+
+```python
+ds.map_batches(Predicator, resources={"TPU": 1})
+```
+
+### Kuberay
+
+To auto populate `rayStartParams` resources using pod resources, some code is needed.
```
Comment (jjyao) [inline reps/2023-10-13-accelerator-support.md:213]: @kevin85421 for review kuberay section.

**Review by kevin85421 — APPROVED** [review]
The KubeRay part looks good to me. Some details might be helpful:

* How to schedule multiple bundles in the same placement group to the hosts in a SuperPOD rather than multiple SuperPODs?
* How does autoscaling work for SuperPods?
* How to set up accelerators related environment variables (e.g. TPU_WORKER_ID, TPU_WORKER_HOSTNAMES)?

See https://github.com/ray-project/ray/issues/39781#issuecomment-1743658075 for more details.

File: reps/2023-10-13-accelerator-support.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+Besides defining the accelerator subclass, the accelerator resource needs to be marked as unit instance resource in config `RAY_custom_unit_instance_resources`.
+
+### Ray Train
+TODO
+
+### Ray RLLib
+TODO
+
+### Ray Serve
```
Comment (jjyao) [inline reps/2023-10-13-accelerator-support.md:187]: @edoakes for reviewing the serve section.

File: reps/2023-10-13-accelerator-support.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+@serve.deployment(ray_actor_options={"num_gpus": 1})
+class Deployment:
+    ...
+
+@serve.deployment(ray_actor_options={"resources": {"TPU": 1}})
+class Deployment:
+    ...
+```
+
+### Ray Data
```
Comment (jjyao) [inline reps/2023-10-13-accelerator-support.md:202]: @raulchen for reviewing the data section.

File: reps/2023-10-13-accelerator-support.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+## Summary
+
+Holistic design of supporting accelerators in Ray.
+This also serves as the reference doc for contributors
+to add new accelerator support in the future.
+
+### General Motivation
+
+Nowadays, more and more specialized accelerators (e.g. TPU, HPU) are coming out to speed up AI workloads.
+Supporting those accelerators natively in Ray will increase Ray's adoption as the computing framework for scaling AI applications.
```
Comment (edoakes) [inline reps/2023-10-13-accelerator-support.md:12]: We should define what "supporting them natively" means at a high level. i.e., what does native support enable?

File: reps/2023-10-13-accelerator-support.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+@ray.remote(resources={"TPU": 1})
+def task():
+    ...
+
+@ray.remote(resources={"TPU": 1}, accelerator_type="TPU-V4")
+def task():
+    ...
+```
+
+##### Alternatives
```
Comment (edoakes) [inline reps/2023-10-13-accelerator-support.md:59]: For the alternatives, can you briefly describe why we went with the above instead of them?

File: reps/2023-10-13-accelerator-support.md [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    ...
+```
+
+#### Implementation
+
+Core defines an `Accelerator` base class and each accelerator needs to implement a subclass (e.g. `TPUAccelerator`):
+
+```python
+@DeveloperAPI
+class Accelerator(ABC):
```
Comment (edoakes) [inline reps/2023-10-13-accelerator-support.md:113]: This class name is sort of odd given -- from the name I assumed it represented/managed a single physical accelerator, but it seems to instead define various utilities for interacting with all accelerators on a node. I'd suggest something like `AcceleratorManager` or `AcceleratorManagementInterface`

File: reps/2023-10-13-accelerator-support.md [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        """Get the total number of accelerators on this node."""
+
+    @staticmethod
+    @abstractmethod
+    def get_visible_accelerator_ids() -> Optional[List[str]]:
+        """Get the ids of accelerators visible to Ray."""
+
+    @staticmethod
+    @abstractmethod
+    def get_accelerator_type() -> Optional[str]:
```
Comment (edoakes) [inline reps/2023-10-13-accelerator-support.md:141]: how's this different from `get_resource_name`?

File: reps/2023-10-13-accelerator-support.md [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+#### Implementation
+
+Core defines an `Accelerator` base class and each accelerator needs to implement a subclass (e.g. `TPUAccelerator`):
+
+```python
+@DeveloperAPI
+class Accelerator(ABC):
+    @staticmethod
+    @abstractmethod
+    def is_available() -> bool:
```
Comment (edoakes) [inline reps/2023-10-13-accelerator-support.md:116]:
```suggestion
    def is_available_on_current_node() -> bool:
```

Otherwise, by default i'd assume this is checking if the accelerator is available on the cluster.

File: reps/2023-10-13-accelerator-support.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        """Get the type of the accelerator on this node.
+
+        The result should only be used when get_num_accelerators() > 0.
+
+        Return None if it's unknown.
+        """
+
+    @staticmethod
+    @abstractmethod
+    def validate_resource_request_quantity(quantity: float) -> Optional[str]:
```
Comment (edoakes) [inline reps/2023-10-13-accelerator-support.md:151]: why's the return value optional?

File: reps/2023-10-13-accelerator-support.md (hunk 3 above)
Comment (edoakes) [reply reps/2023-10-13-accelerator-support.md:187]: LGTM

**Review by edoakes — COMMENTED** [review]
Looks like a good proposal overall. Some general comments on the API. Serve part LGTM.

File: reps/2023-10-13-accelerator-support.md (hunk 8 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:141]: `get_resource_name()` returns "TPU" but `get_accelerator_type()` returns "TPU-V4", "TPU-V5" etc.

File: reps/2023-10-13-accelerator-support.md (hunk 10 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:151]: This returns the error message if validation fails. Return None if it's valid.

File: reps/2023-10-13-accelerator-support.md (hunk 5 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:12]: I'll add more details. But basically the same support level as existing Nvidia GPU.

File: reps/2023-10-13-accelerator-support.md [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+    ...
+
+@ray.remote(num_accelerators=1, accelerator_type="TPU-V4")
+def task()
+    ...
+```
+
+#### Implementation
+
+Core defines an `Accelerator` base class and each accelerator needs to implement a subclass (e.g. `TPUAccelerator`):
```
Comment (matthewdeng) [inline reps/2023-10-13-accelerator-support.md:109]: A few questions that I wasn't sure if we needed to clarify in the implementation section:

1. Anything to add for accelerator types here?
2. Do we need to clarify between virtual and physical resources?
3. Maybe out of scope for this REP, but do we need to consider partial resources?

File: reps/2023-10-13-accelerator-support.md [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+## Design
+
+### Ray Core
+
+#### API
+Users will use accelerators through Ray resources. For GPU accelerators (e.g. Nvidia, AMD, Intel GPUs), they will use the `GPU` resource name. For other accelerators, they will each have their own resource name (e.g. TPU accelerator has resource name `TPU`). `accelerator_type` option can also be used to specify a particular type of accelerator for a task or actor.
+
+```python
+@ray.remote(num_gpus=1)
```
[the remark this answers is not in the document]
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:38]: Yes. Although you have to use the `num_gpus` syntax sugar.

File: reps/2023-10-13-accelerator-support.md (hunk 7 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:113]: I got the naming from https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/accelerators/accelerator.py.  But if people think its confusing, we can change to `AcceleratorManager`.

File: reps/2023-10-13-accelerator-support.md [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        """
+        return None
+
+    @staticmethod
+    def get_ec2_accelerator_type(instance_type: str, instances: dict) -> Optional[str]:
+        """Get the accelerator type of this aws ec2 instance type.
+
+        Return None if it's unknown.
+        """
+        return None
```
[the remark this answers is not in the document]
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:176]: Ideally yes but currently only aws node provider supports this.

File: reps/2023-10-13-accelerator-support.md (hunk 11 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:109]: > Anything to add for accelerator types here?

You need to implement the `get_accelerator_type()` abstract method.

> Do we need to clarify between virtual and physical resources?

Here it's for auto detecting physical resources. 

> Maybe out of scope for this REP, but do we need to consider partial resources?

Partial resources are supported by Ray out of the box. We don't need to do anything here. But if the accelerator doesn't support fractional, you should return error in `validate_resource_request_quantity`

File: reps/2023-10-13-accelerator-support.md (hunk 10 above)
Comment (edoakes) [reply reps/2023-10-13-accelerator-support.md:151]: oh... maybe return a `Tuple[bool, str]` with `(valid, msg)`?

File: reps/2023-10-13-accelerator-support.md [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+
+@ray.remote(num_gpus=1, accelerator_type="Intel-Max-1550")
+def task():
+    ...
+
+@ray.remote(resources={"TPU": 1})
+def task():
+    ...
+
+@ray.remote(resources={"TPU": 1}, accelerator_type="TPU-V4")
```
Comment (woshiyyya) [inline reps/2023-10-13-accelerator-support.md:54]: Who defines these accelerator types? Should it be defined by the user, or do we detect it automatically?

File: reps/2023-10-13-accelerator-support.md [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@scv119
+
+## Design
+
+### Ray Core
+
+#### API
+Users will use accelerators through Ray resources. For GPU accelerators (e.g. Nvidia, AMD, Intel GPUs), they will use the `GPU` resource name. For other accelerators, they will each have their own resource name (e.g. TPU accelerator has resource name `TPU`). `accelerator_type` option can also be used to specify a particular type of accelerator for a task or actor.
```
Comment (woshiyyya) [inline reps/2023-10-13-accelerator-support.md:35]: Are we going to maintain a mapping from accelerator type to resource name?

Do we have something like
```python
ACCELERATOR_MAPPING = {
   "A100": "GPU",
   "Intel-Max-1550": "GPU",
   "TPU-V4": "TPU",
   "Trainium": "neuron_cores",
}
```

File: reps/2023-10-13-accelerator-support.md (hunk 4 above)
Comment (raulchen) [reply reps/2023-10-13-accelerator-support.md:202]: > `ds.map_batches(Predicator, resources={"TPU": 1})`

The correct syntax should be `ds.map_batches(..., ray_remote_args={"resources": {"TPU": 1}})`
Also, we've deprecated the `Predicator` API. So better not to add it in the example. 
Otherwise looks good.

File: reps/2023-10-13-accelerator-support.md [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+        """Get the accelerator type of this aws ec2 instance type.
+
+        Return None if it's unknown.
+        """
+        return None
+```
+
+Besides defining the accelerator subclass, the accelerator resource needs to be marked as unit instance resource in config `RAY_custom_unit_instance_resources`.
+
+### Ray Train
```
Comment (woshiyyya) [inline reps/2023-10-13-accelerator-support.md:181]: @matthewdeng Can you review this and we can discuss offline.

File: reps/2023-10-13-accelerator-support.md (hunk 7 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:113]: Changed to `AcceleratorManager`.

File: reps/2023-10-13-accelerator-support.md (hunk 8 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:141]: Added comments to explain this.

File: reps/2023-10-13-accelerator-support.md (hunk 10 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:151]: Changed.

File: reps/2023-10-13-accelerator-support.md [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+
+    @staticmethod
+    @abstractmethod
+    def get_resource_name() -> str:
+        """Get the name of the resource representing this accelerator."""
+
+    @staticmethod
+    @abstractmethod
+    def get_visible_accelerator_ids_env_var() -> str:
+        """Get the env var that sets the ids of visible accelerators."""
```
[the remark this answers is not in the document]
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md]: Yes.

File: reps/2023-10-13-accelerator-support.md [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+
+    @staticmethod
+    @abstractmethod
+    def get_num_accelerators() -> int:
+        """Get the total number of accelerators on this node."""
+
+    @staticmethod
+    @abstractmethod
+    def get_visible_accelerator_ids() -> Optional[List[str]]:
+        """Get the ids of accelerators visible to Ray."""
```
[the remark this answers is not in the document]
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md]: Renamed to `get_current_process_visible_accelerator_ids`

File: reps/2023-10-13-accelerator-support.md (hunk 14 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:54]: It's auto detected.

File: reps/2023-10-13-accelerator-support.md (hunk 15 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:35]: No. Is there a use case for that?

File: reps/2023-10-13-accelerator-support.md (hunk 6 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:59]: Added.

File: reps/2023-10-13-accelerator-support.md [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+#### 1. Specify Accelerator Type in `ScalingConfig`
+
+Ray Core automatically detects the accelerator type, and treats it as an additional resource tag. Ray Train will therefore be able to launch training jobs on specific accelerator types. We add a new initialization argument `accelerator_type` for `ray.train.ScalingConfig`.
+
+
+```python
+# CPU
+ScalingConfig(num_workers=4)
+
+# GPU
```
Comment (jjyao) [inline reps/2023-10-13-accelerator-support.md:253]: @woshiyyya  Can you add an example of Intel GPU?

File: reps/2023-10-13-accelerator-support.md [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+
+Internally, Ray Train adds the `accelerate_type` parameter in `ray.remote()` to start the training workers.
+
+#### 2. Specify PyTorch Backend for each Accelerator Family
+
+Different accelerator family uses different backend to launch deep learning training. Ray Train will implement a different backend class for each accelerator family to launch distributed groups and set environment variables.
+
+| Accelerator | Communication Backend | Developer APIs | User APIs | 
+| - | - | - | - |
+| GPU      | `nccl`, `gloo` | TorchBackend | `TorchConfig()` |
```
Comment (jjyao) [inline reps/2023-10-13-accelerator-support.md:279]: @woshiyyya Can you add an example of Intel GPU

Comment (woshiyyya) [reply reps/2023-10-13-accelerator-support.md:275]: Done.

File: reps/2023-10-13-accelerator-support.md (hunk 15 above)
Comment (scv119) [reply reps/2023-10-13-accelerator-support.md:35]: let's document the mapping in a table in this doc?
"GPU" -> Nvidia, Intel, AMD
"TPU" -> TPU
...

File: reps/2023-10-13-accelerator-support.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (scv119) [reply reps/2023-10-13-accelerator-support.md:38]: if we use `num_gpus` and not specify accelerators, the train or other library might get a intel/amd gpu. How would  the worker detect that and work with different type of gpus?

File: reps/2023-10-13-accelerator-support.md (hunk 15 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md]: Added.

File: reps/2023-10-13-accelerator-support.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:48]: So on each node you only have intel or amd gpus and we set the correct visible accelerator ids env var based on what the node has physically.

**Comment by richardliaw** [comment]
@sriram-anyscale , would be great to get your thoughts on this as well.

File: reps/2023-10-13-accelerator-support.md [hunk 21]
```diff
@@ -0,0 +1,399 @@
...
+| ...         | ...                       |
+
+`accelerator_type` option can also be used to specify a particular type of accelerator for a task or actor.
+
+```python
+@ray.remote(num_gpus=1)
+def task():
+    ...
+
+@ray.remote(num_gpus=1, accelerator_type="A100")
```
Comment (sriram-anyscale) [inline reps/2023-10-13-accelerator-support.md:52]: How would say that you are OK with either A100 or A10G?

File: reps/2023-10-13-accelerator-support.md [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+@scv119
+
+## Design
+
+### Ray Core
+
+#### API
+Users will use accelerators through Ray resources. For GPU accelerators (e.g. Nvidia, AMD, Intel GPUs), they will use the `GPU` resource name. For other accelerators, they will each have their own resource name (e.g. TPU accelerator has resource name `TPU`).
+
+| Resource    | Accelerator               |
```
Comment (sriram-anyscale) [inline reps/2023-10-13-accelerator-support.md:37]: Is the distinction between GPU, TPU, HPU, etc. important? Can't we use a more generic term like "num_accelerators"? It also seems a bit inconsistent to have "num_gpus", but not "num_tpus", etc. - I'm guessing it is because you don't want too many of those "num_*" attributes.

You do mention that "num_accelerators" was an alternative you considered - but discarded it. The alternative you have chosen to special case GPUs with num_gpus but not for the others seems like it could do with a bit more cleanup.

File: reps/2023-10-13-accelerator-support.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (sriram-anyscale) [reply reps/2023-10-13-accelerator-support.md:48]: I've asked a similar question to Chen below - essentially you can say you want a specific accelerator, or you don't care which accelerator you get. But you cannot say that you only want one of two accelerators (as in my example in the other comment - "I am OK with either of A100 or A10G").

File: reps/2023-10-13-accelerator-support.md [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+    ...
+
+@ray.remote(num_tpus=1, accelerator_type="TPU-V4")
+def task():
+    ...
+```
+
+A single `ACCELERATOR` resource for the ability to schedule a task on any accelerator
+and an `accelerator_family` is also introduced to allow specifying the family of the accelerators to use.
+People feel this is less obvious or expclit than having a reosurce name per accelerator family
```
Comment (sriram-anyscale) [inline reps/2023-10-13-accelerator-support.md:88]: typo (reosurce)

File: reps/2023-10-13-accelerator-support.md (hunk 22 above)
Comment (sriram-anyscale) [reply reps/2023-10-13-accelerator-support.md:37]: The H100 datasheet says this:

Note that the H100 GPUs are primarily built for executing
datacenter and edge compute workloads for AI, HPC, and data analytics, but not graphics
processing.

So blurs the distinction between GPU and other AI processors.

File: reps/2023-10-13-accelerator-support.md (hunk 21 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:52]: accelerator_type is just labels/custom resources that Ray adds for you automatically. For your case, you can just manually create your own label/custom resource like `A100_OR_A10G` and request it in your task.

File: reps/2023-10-13-accelerator-support.md (hunk 22 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:37]: We try to have the resource name match how these accelerators are being called by people.

> Is the distinction between GPU, TPU, HPU, etc. important?

Different accelerators are quite different: e.g. some supports fractional resource, some not. Also the user's code normally only works for one accelerator family since different accelerator family provides different programming model (e.g. Nvidia GPU uses cuda).
This means if we do `num_accelerators`, we also need to introduce an `accelerator_family` node label so uses can express the equivalent of `num_gpus=1` with `num_accelerators=1, accelerator_family="GPU"`

File: reps/2023-10-13-accelerator-support.md (hunk 21 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:52]: Later on, once we switch to label you can say `"accelerator_type": In("A100", "A10G")`

File: reps/2023-10-13-accelerator-support.md [hunk 24]
```diff
@@ -0,0 +1,399 @@
...
+
+#### API
+Users will use accelerators through Ray resources. For GPU accelerators (e.g. Nvidia, AMD, Intel GPUs), they will use the `GPU` resource name. For other accelerators, they will each have their own resource name (e.g. TPU accelerator has resource name `TPU`).
+
+| Resource    | Accelerator               |
+| ----------- | ------------------------- |
+| GPU         | Nvidia, Intel, AMD, Apple |
+| TPU         | TPU                       |
+| HPU         | Habana Gaudi              |
+| neuron_cores| AWS Trainium, Inferentia  |
```
Comment (ArturNiederfahrenhorst) [inline reps/2023-10-13-accelerator-support.md:42]: Can someone ELI5 what the difference between TPU, GPU and neuron_cores is? And what other _types_ of accelerators might be part of the triple dots?

My angle is, if these all do matrix multiplications, from a user perspective, why would I need to change my codebase to use another of these resources?

Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:42]: Different accelerators have different programming models. E.g. Nvidia GPU uses cuda and TPU uses XLA so normally you need to change your codebase when you switch from one type of accelerator to another.

File: reps/2023-10-13-accelerator-support.md (hunk 21 above)
Comment (rkooo567) [reply reps/2023-10-13-accelerator-support.md:52]: Maybe we can allow syntactic sugar like accelerator_type="A100,A10"? Feel like this is more pythonic API

File: reps/2023-10-13-accelerator-support.md [hunk 25]
```diff
@@ -0,0 +1,399 @@
...
+```
+
+### Kuberay
+
+To auto populate `rayStartParams` resources using pod resources, some code is needed.
+See https://github.com/ray-project/kuberay/blob/b7bc7ae8c983160bb700d6b8bb4014dda183ecea/ray-operator/controllers/ray/common/pod.go#L687 as an example of how GPU resource is populated.
+
+### Observability
+
+Currently we emit metrics about the physical usages of GPUs and also show them on Ray dashboard.
```
Comment (rkooo567) [inline reps/2023-10-13-accelerator-support.md:382]: Can we make this as a part of `AcceleratorManager` class? We also need ways to emit metrics. 

We need: acclerator (TPU vs GPU), acclerator_type (A100, H100), utilization & memory. 

1. Dashboard cluster view should display live metrics.
2. Dashboard should emit metrics with device type (with different labels). 
3. ray status should correctly display devices.

File: reps/2023-10-13-accelerator-support.md (hunk 21 above)
Comment (akshay-anyscale) [reply reps/2023-10-13-accelerator-support.md:52]: there's also the proposed placement group alternatives feature which may be a better way of solving that

File: reps/2023-10-13-accelerator-support.md (hunk 25 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:382]: Yes, it will. I'm trying some prototype here: https://github.com/ray-project/ray/pull/40749.

File: reps/2023-10-13-accelerator-support.md [hunk 26]
```diff
@@ -0,0 +1,399 @@
...
+  def get_resource_name() -> str:
+      """Get the name of the resource representing this accelerator family.
+
+      Returns:
+          The resource name: e.g., the resource name for Nvidia GPUs is "GPU"
+      """
+
+  @staticmethod
+  @abstractmethod
+  def get_visible_accelerator_ids_env_var() -> str:
```
Comment (ericl) [inline reps/2023-10-13-accelerator-support.md:159]: Do we still need this given that get_current_process_visible_accelerator_ids and set_current_process_visible_accelerator_ids are available?

File: reps/2023-10-13-accelerator-support.md [hunk 27]
```diff
@@ -0,0 +1,399 @@
...
+    scaling_config=ScalingConfig(
+        num_workers=4, 
+        resources_per_worker={"HPU": 1}
+    ),
+    torch_config=TorchConfig(backend="hccl")
+)
+```
+
+
+### Ray RLLib
```
Comment (ericl) [inline reps/2023-10-13-accelerator-support.md:348]:
```suggestion
### Ray RLlib
```

File: reps/2023-10-13-accelerator-support.md [hunk 28]
```diff
@@ -0,0 +1,399 @@
...
+Currently there is no plan to support other accelerators in RLLib.
```
Comment (ericl) [inline reps/2023-10-13-accelerator-support.md:349]:
```suggestion
Currently there is no plan to support other accelerators in RLlib.
```

File: reps/2023-10-13-accelerator-support.md [hunk 29]
```diff
@@ -0,0 +1,399 @@
...
+  @abstractmethod
+  def set_current_process_visible_accelerator_ids(ids: List[str]) -> None:
+      """Set the ids of accelerators of this family that are visible to the current process.
+
+      Args:
+          ids: The ids of visible accelerators of this family.
+      """
+
+  @staticmethod
+  def get_ec2_instance_num_accelerators(
```
Comment (ericl) [inline reps/2023-10-13-accelerator-support.md:230]: Can we make this cloud agnostic?

File: reps/2023-10-13-accelerator-support.md (hunk 26 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:159]: Currently it's only used in some error messages

```
raise ValueError(
                    f"Attempting to start raylet with {num_accelerators} "
                    f"{accelerator_resource_name}, "
                    f"but {accelerator_manager.get_visible_accelerator_ids_env_var()} "
                    f"contains {visible_accelerator_ids}."
                )
```

File: reps/2023-10-13-accelerator-support.md (hunk 29 above)
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:230]: Not really since this is used by autoscaler to auto fill the resources for each node type before the node is started so for different clouds, there are different ways (mostly just querying metadata for an instance type) to get those information.

**Review by ericl — COMMENTED** [review]
LGTM

File: reps/2023-10-13-accelerator-support.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (anyscalesam) [reply reps/2023-10-13-accelerator-support.md:48]: how would this work with mixed fleets? eg: a ray cluster with both nvidia, amd, and intel custom silicon?

File: reps/2023-10-13-accelerator-support.md (hunk 21 above)
Comment (anyscalesam) [reply reps/2023-10-13-accelerator-support.md:52]: what was the solution/conclusion that we landed on?

File: reps/2023-10-13-accelerator-support.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jjyao) [reply reps/2023-10-13-accelerator-support.md:48]: It should just work.
